unidist/core/backends/mpi/core/controller/api.py

```diff
@@ -368,41 +368,43 @@ def wait(data_ids, num_returns=1):
     """
     mpi_state = communication.MPIState.get_instance()
 
-    def wait_impl(data_id):
-        if object_store.contains(data_id):
-            return
-
-        owner_rank = object_store.get_data_owner(data_id)
-
-        operation_type = common.Operation.WAIT
-        operation_data = {"id": data_id.base_data_id()}
-        # We use a blocking send here because we have to wait for
-        # completion of the communication, which is necessary for the pipeline to continue.
-        communication.send_simple_operation(
-            mpi_state.comm,
-            operation_type,
-            operation_data,
-            owner_rank,
-        )
-
-        logger.debug("WAIT {} id from {} rank".format(data_id._id, owner_rank))
-
-        communication.mpi_busy_wait_recv(mpi_state.comm, owner_rank)
-
+    not_ready = data_ids
+    pending_returns = num_returns
+    ready = []
     logger.debug("WAIT {} ids".format(common.unwrapped_data_ids_list(data_ids)))
+    for data_id in not_ready:
+        if object_store.contains(data_id):
+            ready.append(data_id)
+            not_ready.remove(data_id)
+            pending_returns -= 1
+    if num_returns == len(ready):
+        return ready, not_ready
 
     if not isinstance(data_ids, list):
         data_ids = [data_ids]
+    operation_type = common.Operation.WAIT
+    data_ids = [common.unwrap_data_ids(arg) for arg in data_ids]
+    data = {
+        "operation_type": operation_type,
+        "data_ids": data_ids,
+        "num_returns": pending_returns,
+    }
+    # Monitor the task execution
+    # We use a blocking send and recv here because we have to wait for
+    # completion of the communication, which is necessary for the pipeline to continue.
 
-    ready_count = 0
-    ready = [None] * num_returns
-    not_ready = data_ids
+    communication.mpi_send_object(
+        communication.MPIState.get_instance().comm,
+        data,
+        communication.MPIRank.MONITOR,
+    )
```

Comment:
```suggestion
    communication.send_simple_operation(
        communication.MPIState.get_instance().comm,
        operation_type,
        operation_data,
        communication.MPIRank.MONITOR,
    )
```

Reply:
On monitor side there should be two recvs: 1. recv_operation_type (without modifications), 2. recv_simple_operation in the if branch of wait.